tool/tsh/db.go

```diff
@@ -1067,68 +1055,161 @@ func formatDatabaseConfigCommand(clusterFlag string, db tlsca.RouteToDatabase) s
-// be used for connecting to the provided database.
-func shouldUseLocalProxyForDatabase(tc *client.TeleportClient, db *tlsca.RouteToDatabase) bool {
-	return tc.TLSRoutingEnabled || isLocalProxyAlwaysRequired(tc, db.Protocol)
+// dbLocalProxyRequirement describes local proxy requirements for connecting to a database.
+type dbLocalProxyRequirement struct {
+	// localProxy is whether a local proxy is required to connect.
+	localProxy bool
+	// localProxyReasons is a list of reasons for why local proxy is required.
+	localProxyReasons []string
+	// localProxy is whether a local proxy tunnel is required to connect.
+	tunnel bool
+	// tunnelReasons is a list of reasons for why a tunnel is required.
+	tunnelReasons []string
 }
 
-// isLocalProxyAlwaysRequired returns true for protocols that always requires
-// an ALPN local proxy.
-func isLocalProxyAlwaysRequired(tc *client.TeleportClient, protocol string) bool {
+// requireLocalProxy sets the local proxy requirement and appends reasons.
+func (r *dbLocalProxyRequirement) requireLocalProxy(reasons ...string) {
+	r.localProxy = true
+	r.localProxyReasons = append(r.localProxyReasons, reasons...)
+}
+
+// requireLocalProxyWithTunnel sets the local proxy and tunnel requirements,
+// and appends reasons for both.
+func (r *dbLocalProxyRequirement) requireLocalProxyWithTunnel(reasons ...string) {
+	r.requireLocalProxy(reasons...)
+	r.tunnel = true
+	r.tunnelReasons = append(r.tunnelReasons, reasons...)
+}
+
+// getDBLocalProxyRequirement determines what local proxy settings are required
+// for a given database.
+func getDBLocalProxyRequirement(tc *client.TeleportClient, route *tlsca.RouteToDatabase) *dbLocalProxyRequirement {
+	var req dbLocalProxyRequirement
 	switch tc.PrivateKeyPolicy {
 	case keys.PrivateKeyPolicyHardwareKey, keys.PrivateKeyPolicyHardwareKeyTouch:
-		return true
-	}
-	switch protocol {
-	case defaults.ProtocolSQLServer,
-		defaults.ProtocolSnowflake,
-		defaults.ProtocolCassandra,
-		defaults.ProtocolDynamoDB:
-		return true
-	default:
-		return false
+		req.requireLocalProxyWithTunnel(formatKeyPolicyReason(tc.PrivateKeyPolicy))
+	}
+
+	// Some protocols (Snowflake, DynamoDB) only works in the local tunnel mode.
+	switch route.Protocol {
+	case defaults.ProtocolSnowflake, defaults.ProtocolDynamoDB:
+		req.requireLocalProxyWithTunnel(formatDBProtocolReason(route.Protocol))
+	case defaults.ProtocolSQLServer, defaults.ProtocolCassandra:
+		req.requireLocalProxy(formatDBProtocolReason(route.Protocol))
+	case defaults.ProtocolMySQL:
+		if tc.TLSRoutingEnabled {
+			req.requireLocalProxy(fmt.Sprintf("%v and %v",
+				formatDBProtocolReason(route.Protocol),
+				formatTLSRoutingReason(tc.SiteName)))
+		}
 	}
+	return &req
+}
+
+// dbLocalProxyPreference contains local proxy preferences for connecting to a database,
+// i.e. `tsh db connect` will connect using preferred local proxy settings.
+type dbLocalProxyPreference struct {
```

Comment:
nit: the two structs seem identical. Would it be easier to combine them? Or you prefer the clarity of having both?

Reply:
I originally had them as one struct. I separated them to create a distinction between what is required and what is "preferred". I also didn't want `tsh db config` or `tsh db env` to be doing preference checks rather than required checks, so it just seemed better to split them up

Reply:
refactored to remove this actually, after consider this and your other comment